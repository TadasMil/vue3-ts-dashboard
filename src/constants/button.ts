export const ButtonColors = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',
  HIGHLIGHT: 'highlight'
} as const

export const ButtonDensities = {
  DEFAULT: 'default',
  COMFORTABLE: 'comfortable',
  COMPACT: 'compact',
  NONE: 'none'
} as const

export const ButtonVariants = {
  DEFAULT: 'default',
  OUTLINED: 'outlined',
  TEXT: 'text'
} as const

export const ButtonRoundeds = {
  SM: 'sm',
  MD: 'md',
  LG: 'lg'
} as const
