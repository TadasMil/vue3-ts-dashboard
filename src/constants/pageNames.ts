export default {
  /* User Related Pages */
  UserDashboard: 'UserDashboard',
  UserRecruitment: 'UserRecruitment',
  UserSchedule: 'UserSchedule',
  UserSupport: 'UserSupport',
  UserSettings: 'UserSettings',
  UserDepartment: 'UserDepartment',
  UserEmployee: 'UserEmployee'
}
