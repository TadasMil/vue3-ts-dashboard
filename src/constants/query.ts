export const QUERY_STATES = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  ERROR: 'ERROR',
  NOT_FOUND: 'NOT_FOUND',
  NO_RESULTS: 'NO_RESULTS',
  NOT_AUTHORIZED: 'NOT_AUTHORIZED'
} as const
