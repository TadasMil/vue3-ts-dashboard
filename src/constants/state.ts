export const STATES = {
  LOADING: 'LOADING',
  IDLE: 'IDLE',
  ERROR: 'ERROR',
  NOT_FOUND: 'NOT_FOUND',
  NO_RESULTS: 'NO_RESULTS',
  NOT_AUTHORIZED: 'NOT_AUTHORIZED'
} as const
