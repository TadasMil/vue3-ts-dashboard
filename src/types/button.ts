import { ButtonColors, ButtonDensities, ButtonRoundeds, ButtonVariants } from '@/constants/button'

export type ButtonColor = (typeof ButtonColors)[keyof typeof ButtonColors]
export type ButtonDensity = (typeof ButtonDensities)[keyof typeof ButtonDensities]
export type ButtonVariant = (typeof ButtonVariants)[keyof typeof ButtonVariants]
export type ButtonRounded = (typeof ButtonRoundeds)[keyof typeof ButtonRoundeds]
