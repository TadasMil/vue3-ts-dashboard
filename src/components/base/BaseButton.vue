<script setup lang="ts" name="BaseButton">
import { computed } from 'vue'
import { RouteLocationRaw } from 'vue-router'

import BaseIcon from '@/components/base/BaseIcon.vue'

import { ButtonColor, ButtonDensity, ButtonRounded, ButtonVariant } from '@/types/button'
import { IconType } from '@/types/icons'

const {
  color = 'primary',
  icon,
  density = 'default',
  disabled,
  size,
  variant = 'default',
  href,
  to,
  rounded = 'sm'
} = defineProps<{
  icon?: IconType
  size?: number
  disabled?: boolean
  color?: ButtonColor
  density?: ButtonDensity
  variant?: ButtonVariant
  rounded?: ButtonRounded
  href?: string
  to?: RouteLocationRaw
}>()

const emit = defineEmits<{
  (e: 'click'): void
}>()

const tag = computed(() => {
  if (to) return 'router-link'
  if (href) return 'a'
  return 'button'
})

const classes = computed(() => [
  `btn-${color}`,
  `btn-density-${density}`,
  `btn-rounded-${rounded}`,
  `btn-variant-${icon ? 'icon' : variant ?? 'default'}`,
  { 'btn-inverted': icon },
  { 'btn-disabled': disabled }
])

const styles = computed(() => ({
  width: `${size}px`,
  height: `${size}px`
}))
</script>

<template>
  <component
    v-bind="$attrs"
    :is="tag"
    :to="to"
    :href="href"
    type="submit"
    :disabled="disabled"
    :class="[...classes, 'btn']"
    :style="styles"
    @click="emit('click')"
  >
    <BaseIcon v-if="icon" :icon="icon" :size="size" />
    <slot v-else />
  </component>
</template>

<style scoped>
.btn-primary {
  --btn-background: var(--color-primary);
  --btn-foreground: var(--color-white);
  --btn-background-hover: rgba(255, 151, 151, 0.2);
  --btn-background-active: 2px solid var(--color-primary-light);

  --btn-outlined-border: #ff9797;
  --btn-outlined-background: var(--color-white);
  --btn-outlined-background-hover: rgba(255, 151, 151, 0.2);
  --btn-outlined-active-border: var(--color-primary);

  --inverted-btn-foreground: var(--color-primary);
  --inverted-btn-foreground-hover: var(--color-primary-dark);
}

.btn-highlight {
  --btn-background: var(--color-highlight);
  --btn-foreground: var(--color-white);
  --btn-background-hover: rgba(178, 178, 178, 0.2);
  --btn-background-active: 2px solid var(--color-highlight-light);

  --btn-outlined-border: #989898;
  --btn-outlined-background: var(--color-white);
  --btn-outlined-background-hover: rgba(178, 178, 178, 0.2);
  --btn-outlined-active-border: var(--color-highlight);

  --inverted-btn-foreground: var(--color-highlight);
  --inverted-btn-foreground-hover: var(--color-highlight-dark);
}

/* 
TODO: Define styles for the remaining button colors from the color palette

.btn-tertiary {}
*/

.btn {
  background-color: var(--btn-background);
  color: var(--btn-foreground);
  border: 1px solid var(--btn-background);
  transition:
    background-color 0.3s,
    color 0.3s;
}

.btn.btn-variant-default {
  background-color: var(--btn-background);
  color: var(--btn-foreground);
}

.btn.btn-variant-default:hover {
  box-shadow: 0px 4px 12px 0px var(--btn-background-hover);
}

.btn.btn-variant-default:active {
  border: var(--btn-background-active);
}

.btn.btn-variant-outlined {
  color: var(--btn-background);
  border: 1px solid var(--btn-outlined-border);
  background-color: var(--btn-outlined-background);
}

.btn.btn-variant-outlined:hover {
  background-color: var(--btn-outlined-background-hover);
}

.btn.btn-variant-outlined:active {
  border: 1px solid var(--btn-outlined-active-border);
}

.btn.btn-variant-text {
  border: none;
  background-color: transparent;
  color: var(--inverted-btn-foreground);
}

.btn.btn-variant-text:hover {
  color: var(--inverted-btn-foreground-hover);
}

.btn.btn-inverted {
  border: none;
  background-color: transparent;
  color: var(--inverted-btn-foreground);
}

.btn.btn-inverted:hover {
  color: var(--inverted-btn-foreground-hover);
}

.btn-density-default {
  padding: 10px 15px;
}

.btn-density-comfortable {
  padding: 12px 18px;
}

.btn-density-compact {
  padding: 8px 12px;
}

.btn-density-none {
  padding: 0;
}

.btn-disabled {
  opacity: 0.6;
  pointer-events: none;
}

.btn-rounded-sm {
  border-radius: 4px;
}

.btn-rounded-md {
  border-radius: 8px;
}

.btn-rounded-lg {
  border-radius: 16px;
}
</style>
