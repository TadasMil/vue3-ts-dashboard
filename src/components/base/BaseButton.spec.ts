import { h } from 'vue'
import { fireEvent, render } from '@testing-library/vue'
import { describe, expect, it } from 'vitest'

import { ButtonColors, ButtonDensities, ButtonRoundeds, ButtonVariants } from '@/constants/button'

import BaseButton from './BaseButton.vue'

const colorCases = Object.values(ButtonColors)
const densityCases = Object.values(ButtonDensities)
const variantCases = Object.values(ButtonVariants)
const roundedCases = Object.values(ButtonRoundeds)

describe('BaseButton.vue', () => {
  colorCases.forEach((color) => {
    it(`should render button with appropriate color of ${color}`, () => {
      const { html } = render(BaseButton, {
        props: {
          color
        }
      })

      expect(html()).toMatchSnapshot()
    })

    it(`should render button with appropriate color of ${color} and disabled`, () => {
      const { html } = render(BaseButton, {
        props: {
          color,
          disabled: true
        }
      })

      expect(html()).toMatchSnapshot()
    })
  })

  densityCases.forEach((density) => {
    it(`should render button with appropriate density of ${density}`, () => {
      const { html } = render(BaseButton, {
        props: {
          density
        }
      })

      expect(html()).toMatchSnapshot()
    })
  })

  variantCases.forEach((variant) => {
    it(`should render button with appropriate variant of ${variant}`, () => {
      const { html } = render(BaseButton, {
        props: {
          variant
        }
      })

      expect(html()).toMatchSnapshot()
    })

    it(`should render button with appropriate variant of ${variant} and disabled`, () => {
      const { html } = render(BaseButton, {
        props: {
          variant,
          disabled: true
        }
      })

      expect(html()).toMatchSnapshot()
    })
  })

  roundedCases.forEach((rounded) => {
    it(`should render button with appropriate rounded of ${rounded}`, () => {
      const { html } = render(BaseButton, {
        props: {
          rounded
        }
      })

      expect(html()).toMatchSnapshot()
    })
  })

  it('should render icon button', () => {
    const { html } = render(BaseButton, {
      props: {
        icon: () => h('div')
      }
    })

    expect(html()).toMatchSnapshot()
  })

  it('should emit click event', async () => {
    const { getByRole, emitted } = render(BaseButton)

    const button = getByRole('button')

    await fireEvent.click(button)

    expect(emitted()).toHaveProperty('click')
  })

  it('should render router-link component', () => {
    const { html } = render(BaseButton, {
      props: {
        to: {
          name: 'MyPath'
        }
      }
    })

    expect(html()).toMatchSnapshot()
  })

  it('should render anchor component', () => {
    const { html } = render(BaseButton, {
      props: {
        href: '#'
      }
    })

    expect(html()).toMatchSnapshot()
  })
})
