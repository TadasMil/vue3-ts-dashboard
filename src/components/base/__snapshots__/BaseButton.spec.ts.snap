// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseButton.vue > should render anchor component 1`] = `"<a data-v-33235260=\\"\\" href=\\"#\\" type=\\"submit\\" disabled=\\"false\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-default btn\\" role=\\"link\\" aria-disabled=\\"false\\"></a>"`;

exports[`BaseButton.vue > should render button with appropriate color of highlight 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-highlight btn-density-default btn-rounded-sm btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate color of highlight and disabled 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" disabled=\\"\\" class=\\"btn-highlight btn-density-default btn-rounded-sm btn-variant-default btn-disabled btn\\" role=\\"button\\" aria-disabled=\\"true\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate color of primary 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate color of primary and disabled 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" disabled=\\"\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-default btn-disabled btn\\" role=\\"button\\" aria-disabled=\\"true\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate color of secondary 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-secondary btn-density-default btn-rounded-sm btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate color of secondary and disabled 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" disabled=\\"\\" class=\\"btn-secondary btn-density-default btn-rounded-sm btn-variant-default btn-disabled btn\\" role=\\"button\\" aria-disabled=\\"true\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate color of tertiary 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-tertiary btn-density-default btn-rounded-sm btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate color of tertiary and disabled 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" disabled=\\"\\" class=\\"btn-tertiary btn-density-default btn-rounded-sm btn-variant-default btn-disabled btn\\" role=\\"button\\" aria-disabled=\\"true\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate density of comfortable 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-comfortable btn-rounded-sm btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate density of compact 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-compact btn-rounded-sm btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate density of default 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate density of none 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-none btn-rounded-sm btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate rounded of lg 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-default btn-rounded-lg btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate rounded of md 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-default btn-rounded-md btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate rounded of sm 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate variant of default 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-default btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate variant of default and disabled 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" disabled=\\"\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-default btn-disabled btn\\" role=\\"button\\" aria-disabled=\\"true\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate variant of outlined 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-outlined btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate variant of outlined and disabled 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" disabled=\\"\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-outlined btn-disabled btn\\" role=\\"button\\" aria-disabled=\\"true\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate variant of text 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-text btn\\" role=\\"button\\" aria-disabled=\\"false\\"></button>"`;

exports[`BaseButton.vue > should render button with appropriate variant of text and disabled 1`] = `"<button data-v-33235260=\\"\\" type=\\"submit\\" disabled=\\"\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-text btn-disabled btn\\" role=\\"button\\" aria-disabled=\\"true\\"></button>"`;

exports[`BaseButton.vue > should render icon button 1`] = `
"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-icon btn-inverted btn\\" role=\\"button\\" aria-disabled=\\"false\\">
  <div data-v-33235260=\\"\\" style=\\"width: 24px; height: 24px;\\"></div>
</button>"
`;

exports[`BaseButton.vue > should render icon button with fixed size given size property is provided 1`] = `
"<button data-v-33235260=\\"\\" type=\\"submit\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-icon btn-inverted btn\\" style=\\"width: 24px; height: 24px;\\" role=\\"button\\" aria-disabled=\\"false\\">
  <div data-v-33235260=\\"\\" style=\\"width: 24px; height: 24px;\\"></div>
</button>"
`;

exports[`BaseButton.vue > should render router-link component 1`] = `"<router-link data-v-33235260=\\"\\" to=\\"[object Object]\\" type=\\"submit\\" disabled=\\"false\\" class=\\"btn-primary btn-density-default btn-rounded-sm btn-variant-default btn\\" aria-disabled=\\"false\\"></router-link>"`;
